mlx/ops.cpp

```diff
@@ -158,50 +158,64 @@ array linspace(
       to_stream(s));
 }
 
-array astype(const array& a, Dtype dtype, StreamOrDevice s /* = {} */) {
+array astype(array a, Dtype dtype, StreamOrDevice s /* = {} */) {
   if (dtype == a.dtype()) {
-    return a;
+    return std::move(a);
```

Comment:
I always thought the compiler would do that for you..?

Reply:
This one is a question: previously when you call `astype` and it takes the no-op path, how many copies were there? At most 1 and maybe 0?

`a = astype(a, int32)`

Now how many will there be?  I think 1 unless you do `a = astype(std::move(a), int32)` which we don't.

We use that idiom a lot so I would be careful on this one.. maybe revert it unless you are sure it is actually removing a copy?

Reply:
Yeah for `array astype(const array&)` with the no-op path, there is always 1 copy, because compiler can not convert a `const array&` to `array` without copying it.

With `array astype(array)`, it is still always 1 copy if you pass a lvalue, so things do not become worse. The copy just becomes explicit, which is why I love the pass-by-value style more than the traditional const reference style.